src/Utils/BeanDescriptor.php

```diff
@@ -373,38 +397,68 @@ private function generateBeanConstructor() : MethodGenerator
      */
     private function getDirectForeignKeysDescriptors(): array
     {
+        if ($this->directForeignKeysDescriptors !== null) {
+            return $this->directForeignKeysDescriptors;
+        }
         $fks = $this->tdbmSchemaAnalyzer->getIncomingForeignKeys($this->table->getName());
 
         $descriptors = [];
 
         foreach ($fks as $fk) {
-            $descriptors[] = new DirectForeignKeyMethodDescriptor($fk, $this->table, $this->namingStrategy, $this->annotationParser, $this->beanNamespace);
+            /** @var DirectForeignKeyMethodDescriptor $desc */
+            $desc = $this->checkForDuplicate(new DirectForeignKeyMethodDescriptor($fk, $this->table, $this->namingStrategy, $this->annotationParser, $this->beanNamespace));
+            $descriptors[] = $desc;
         }
 
-        return $descriptors;
+        $this->directForeignKeysDescriptors = $descriptors;
+        return $this->directForeignKeysDescriptors;
     }
 
     /**
      * @return PivotTableMethodsDescriptor[]
      */
     private function getPivotTableDescriptors(): array
     {
+        if ($this->pivotTableDescriptors !== null) {
+            return $this->pivotTableDescriptors;
+        }
         $descs = [];
         foreach ($this->schemaAnalyzer->detectJunctionTables(true) as $table) {
             // There are exactly 2 FKs since this is a pivot table.
             $fks = array_values($table->getForeignKeys());
 
             if ($fks[0]->getForeignTableName() === $this->table->getName()) {
                 list($localFk, $remoteFk) = $fks;
-                $descs[] = new PivotTableMethodsDescriptor($table, $localFk, $remoteFk, $this->namingStrategy, $this->beanNamespace, $this->annotationParser);
+                /** @var PivotTableMethodsDescriptor $desc */
```

Comment:
`checkForDuplicate` could return "void" instead of "self".
That way, you would not need this @var annotation to satisfy PHPStan.